File: leet_code_algorithms/middle_of_the_linked_list/middle_of_the_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# MY SOLUTION
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        node_list = []
        node_list_length = 0
        if head.next == None:
            return head
        current = head
        while current:
            node_list.append(current)
            node_list_length += 1
            current = current.next
        return node_list[node_list_length//2]

# OPTIMAL SOLUTION
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        s,f=head,head
        while (f and f.next):
            s=s.next
            f=f.next.next
        return s

File: leet_code_algorithms/middle_of_the_linked_list/middle_of_the_linked_list.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

const middleNode = (head: ListNode): ListNode | null => {
  let fast: ListNode | null = head;
  let slow: ListNode | null = head;
  while (fast && fast.next) {
    slow = slow ? slow.next : null;
    fast = fast.next.next;
  }
  return slow;
};
